File: JSstudyGroup/5212.js
const readline = require('readline')

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
})

const data = []

rl.on('line', (input) => {
  data.push(input)
}).on('close', () => {
  let [row, col] = data.shift().split(' ').map(Number)

  let map = data.map((i) => {
    return i.split('')
  })

  let [dx, dy] = [
    [-1, 1, 0, 0],
    [0, 0, -1, 1],
  ]

  let waterCord = []

  const turnWater = (x, y) => {
    let count = 0

    for (let i = 0; i < 4; i++) {
      let nx = x + dx[i]
      let ny = y + dy[i]

      if (nx >= 0 && nx < row && ny >= 0 && ny < col) {
        if (map[nx][ny] === 'X') {
          //사각형 밖도 바다여서 그냥 x의 갯수를 세는게 나음
          count++
        }
      }
    }

    if (count < 2) {
      //x의 갯수가 2보다 작으면 바다로 변함
      //x가 최소한 2개는 있어야 땅으로 남기 때문
      return true
    } else {
      return false
    }
  }

  for (let i = 0; i < row; i++) {
    for (let j = 0; j < col; j++) {
      if (map[i][j] !== 'X') continue

      if (turnWater(i, j)) {
        waterCord.push([i, j])
      }
    }
  }

  waterCord.map((i) => {
    map[i[0]][i[1]] = '.'
  })

  //console.log(map)

  let locateX=[]
  let locateY=[]
  for (let i = 0; i < row; i++) {
    for (let j = 0; j < col; j++) {
      if(map[i][j]!=='X') continue


      locateX.push(i)
      locateY.push(j)
    }
  }

  let minRow=Math.min(...locateX)
  let maxRow=Math.max(...locateX)
  let minCol=Math.min(...locateY)
  let maxCol=Math.max(...locateY)
  

  let result=[]
  // for(let i=minRow; i<=maxRow; i++){
  //   result.push([])
  //   for(let j=minCol; j<=maxCol; j++){
  //     result[result.length-1].push(map[i][j])
  //   }
  // }

  for(let i=minRow; i<=maxRow; i++){
    for(let j=minCol; j<=maxCol; j++){
      result.push(map[i][j])
    }
    result.push('\n')
  }

  console.log(result.join(''))

  process.exit()
})

//이걸 그래프이론으로 어떻게 풀지?

//상하좌우에서 3개 이상이 바다면 잠김

// ..........
// ..XX...X..
// .XX.......

// .XX...X
// XX.....

File: JSstudyGroup/5212.py
[n, m] = list(map(int, input().split(" ")))
board = [list(input()) for _ in range(n)]

[dx, dy] = [[-1, 1, 0, 0], [0, 0, -1, 1]]


def check(x, y):
    count = 0

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]

        if nx < 0 or nx > n - 1 or ny < 0 or ny > m - 1:
            count += 1
            continue
        if board[nx][ny] == ".":
            count += 1

    if count >= 3:
        return True
    return False


changeCord = []
landCord = []
for i in range(len(board)):
    for j in range(len(board[i])):
        value = board[i][j]
        if value == "X":
            landCord.append([i, j])
        if value == "X" and check(i, j):
            changeCord.append([i, j])


for i in changeCord:
    [x, y] = i
    board[x][y] = "."


leftCord = list(filter(lambda i: not i in changeCord, landCord))


startRow = sorted(leftCord, key=lambda x: (x[0]))[0][0]
startCol = sorted(leftCord, key=lambda x: (x[1]))[0][1]
endRow = sorted(leftCord, key=lambda x: (-x[0]))[0][0]
endCol = sorted(leftCord, key=lambda x: (-x[1]))[0][1]


result = []
for row in range(startRow, endRow + 1):
    result.append(board[row][startCol : endCol + 1])

print("\n".join(["".join((i)) for i in result]))
